File: app/src/main/java/com/dumuzeyn/mp3player/SoundAnalysisController.java
package com.dumuzeyn.mp3player;

import android.content.Context;
import android.os.Process;
import java.io.Closeable;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

/** Runs one resumable, local-only analysis queue for the current process. */
final class SoundAnalysisController implements Closeable {
    private static final String PREFS = "mp3_player_ui";
    private static final String ENABLED = "soundAnalysisEnabled";
    private static final String CLUSTERING_VERSION = "similarClusteringVersion";
    private static final long CONSTRAINT_RECHECK_MS = 5_000L;
    private final MainActivityCore host;
    private SoundProfileStore store;
    private AudioFeatureExtractor extractor;
    private final SoundClusterEngine clusterEngine = new SoundClusterEngine();
    private final ExecutorService executor = Executors.newSingleThreadExecutor(runnable -> {
        Thread thread = new Thread(runnable, "voltune-sound-analysis");
        thread.setPriority(Thread.MIN_PRIORITY);
        return thread;
    });
    private final AtomicInteger generation = new AtomicInteger();
    private volatile ArrayList<SoundGroup> groups = new ArrayList<>();
    private volatile int total;
    private volatile int analyzed;
    private volatile int failed;
    private volatile int queued;
    private volatile String activeTitle = "";
    private volatile SoundAnalysisConstraints.BlockReason blockReason =
            SoundAnalysisConstraints.BlockReason.NONE;
    private volatile boolean closed;
    private volatile long lastLibrarySignature = Long.MIN_VALUE;
    private boolean updatePosted;

    SoundAnalysisController(MainActivityCore host) {
        this.host = host;
    }

    boolean enabled() {
        return host.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
                .getBoolean(ENABLED, true);
    }

    void toggle() {
        boolean value = !enabled();
        host.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
                .putBoolean(ENABLED, value).apply();
        if (value) {
            lastLibrarySignature = Long.MIN_VALUE;
            onLibraryReady(host.libraryState.tracks);
        } else {
            generation.incrementAndGet();
            activeTitle = "";
            blockReason = SoundAnalysisConstraints.BlockReason.NONE;
            notifyUi();
        }
    }

    String settingLabel() {
        return host.tr("Analyze songs by sound: ", "Анализировать песни по звучанию: ")
                + host.tr(enabled() ? "on" : "off", enabled() ? "вкл" : "выкл");
    }

    void onLibraryReady(List<Track> tracks) {
        if (closed) {
            return;
        }
        long signature = librarySignature(tracks);
        if (signature == lastLibrarySignature) {
            return;
        }
        lastLibrarySignature = signature;
        int requestedGeneration = generation.incrementAndGet();
        ArrayList<Track> snapshot = new ArrayList<>(tracks);
        executor.execute(() -> runQueue(requestedGeneration, snapshot));
    }

    ArrayList<SoundGroup> groups() {
        return new ArrayList<>(groups);
    }

    int total() {
        return total;
    }

    int analyzed() {
        return analyzed;
    }

    int failed() {
        return failed;
    }

    int queued() {
        return queued;
    }

    String activeTitle() {
        return activeTitle;
    }

    SoundAnalysisConstraints.BlockReason blockReason() {
        return blockReason;
    }

    @Override
    public void close() {
        closed = true;
        generation.incrementAndGet();
        executor.shutdownNow();
        if (store != null) {
            store.close();
        }
    }

    private void runQueue(int requestedGeneration, ArrayList<Track> tracks) {
        Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND);
        if (!isCurrent(requestedGeneration)) {
            return;
        }
        if (store == null) {
            store = new SoundProfileStore(host);
            extractor = new AudioFeatureExtractor(host);
        }
        LinkedHashMap<String, TrackAudioProfile> profiles = store.loadProfiles();
        groups = store.loadGroups();
        ArrayList<Track> pending = prepareQueue(tracks, profiles);
        int pendingCount = pending.size();
        boolean clusteringChanged = host.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
                .getInt(CLUSTERING_VERSION, 0) != SoundClusterEngine.CLUSTERING_VERSION;
        store.pruneEmptyGroups();
        groups = store.loadGroups();
        publishCounts(tracks.size(), profiles, pending.size());
        notifyUi();
        if (!enabled()) {
            return;
        }
        if (clusteringChanged && pending.isEmpty() && usableCount(profiles) >= 4) {
            rebuildGroups(profiles);
        }
        for (Track track : pending) {
            if (!runAllowedOrRescheduled(requestedGeneration, tracks)) {
                return;
            }
            activeTitle = track.title;
            store.mark(track, SoundAnalysisState.ANALYZING, "");
            notifyUi();
            try {
                double[] features = extractor.analyze(track,
                        () -> shouldYield(requestedGeneration));
                if (!isCurrent(requestedGeneration)) {
                    return;
                }
                TrackAudioProfile profile = TrackAudioProfile.analyzed(track, features);
                store.saveProfile(profile);
                profiles.put(track.trackId, profile);
                analyzed++;
                queued = Math.max(0, queued - 1);
                updateGroupsAfterProfile(profiles, profile);
            } catch (AudioFeatureExtractor.AnalysisInterruptedException interrupted) {
                if (!isCurrent(requestedGeneration)) {
                    return;
                }
                store.mark(track, SoundAnalysisState.QUEUED, "");
                lastLibrarySignature = Long.MIN_VALUE;
                onLibraryReady(tracks);
                return;
            } catch (Exception error) {
                if (!isCurrent(requestedGeneration)) {
                    return;
                }
                store.mark(track, SoundAnalysisState.FAILED,
                        error.getClass().getSimpleName());
                profiles.put(track.trackId, new TrackAudioProfile(track.trackId,
                        TrackAudioProfile.ANALYSIS_VERSION, track.fileSize, track.lastModified,
                        track.fingerprint, SoundAnalysisState.FAILED, new double[0], "",
                        error.getClass().getSimpleName(), System.currentTimeMillis()));
                failed++;
                queued = Math.max(0, queued - 1);
                VoltuneLog.failure("sound_analysis_failed", error);
            }
            activeTitle = "";
            notifyUi();
        }
        if (pendingCount > 0 || clusteringChanged || groups.isEmpty()) {
            rebuildGroups(profiles);
        }
        activeTitle = "";
        blockReason = SoundAnalysisConstraints.BlockReason.NONE;
        notifyUi();
    }

    private ArrayList<Track> prepareQueue(ArrayList<Track> tracks,
            LinkedHashMap<String, TrackAudioProfile> profiles) {
        ArrayList<Track> pending = new ArrayList<>();
        for (Track track : tracks) {
            TrackAudioProfile profile = profiles.get(track.trackId);
            if (profile != null && profile.matches(track)
                    && (profile.usable() || profile.state == SoundAnalysisState.FAILED)) {
                continue;
            }
            store.mark(track, SoundAnalysisState.QUEUED, "");
            TrackAudioProfile queuedProfile = TrackAudioProfile.pending(
                    track, SoundAnalysisState.QUEUED);
            profiles.put(track.trackId, queuedProfile);
            pending.add(track);
        }
        return pending;
    }

    private void updateGroupsAfterProfile(Map<String, TrackAudioProfile> profiles,
            TrackAudioProfile profile) {
        if (groups.isEmpty() && usableCount(profiles) >= 4) {
            rebuildGroups(profiles);
            return;
        }
        if (!groups.isEmpty()) {
            String groupId = clusterEngine.nearestGroup(profile.features,
                    new ArrayList<>(profiles.values()), groups);
            store.assign(profile.trackId, groupId);
            groups = store.loadGroups();
        }
        if (usableCount(profiles) <= 20) {
            rebuildGroups(profiles);
        }
    }

    private void rebuildGroups(Map<String, TrackAudioProfile> profiles) {
        ArrayList<SoundGroup> rebuilt = clusterEngine.cluster(
                new ArrayList<>(profiles.values()));
        store.replaceGroups(rebuilt);
        groups = store.loadGroups();
        host.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
                .putInt(CLUSTERING_VERSION, SoundClusterEngine.CLUSTERING_VERSION).apply();
    }

    private boolean runAllowedOrRescheduled(int requestedGeneration,
            ArrayList<Track> tracks) {
        if (!isCurrent(requestedGeneration)) return false;
        blockReason = SoundAnalysisConstraints.reason(host, host.isPlaybackPlaying());
        if (blockReason == SoundAnalysisConstraints.BlockReason.NONE) return true;
        activeTitle = "";
        notifyUi();
        host.uiHandler.postDelayed(() -> {
            if (!closed && enabled() && requestedGeneration == generation.get()) {
                lastLibrarySignature = Long.MIN_VALUE;
                onLibraryReady(tracks);
            }
        }, CONSTRAINT_RECHECK_MS);
        return false;
    }

    private boolean shouldYield(int requestedGeneration) {
        return !isCurrent(requestedGeneration)
                || SoundAnalysisConstraints.reason(host, host.isPlaybackPlaying())
                != SoundAnalysisConstraints.BlockReason.NONE;
    }

    private boolean isCurrent(int requestedGeneration) {
        return !closed && enabled() && requestedGeneration == generation.get()
                && !Thread.currentThread().isInterrupted();
    }

    private void publishCounts(int trackCount, Map<String, TrackAudioProfile> profiles,
            int pendingCount) {
        total = trackCount;
        analyzed = 0;
        failed = 0;
        for (TrackAudioProfile profile : profiles.values()) {
            if (profile.usable()) {
                analyzed++;
            } else if (profile.state == SoundAnalysisState.FAILED) {
                failed++;
            }
        }
        queued = pendingCount;
    }

    private static int usableCount(Map<String, TrackAudioProfile> profiles) {
        int result = 0;
        for (TrackAudioProfile profile : profiles.values()) {
            if (profile.usable()) {
                result++;
            }
        }
        return result;
    }

    private static long librarySignature(List<Track> tracks) {
        long value = 0xcbf29ce484222325L;
        for (Track track : tracks) {
            value = (value ^ track.trackId.hashCode()) * 0x100000001b3L;
            value = (value ^ track.fileSize) * 0x100000001b3L;
            value = (value ^ track.lastModified) * 0x100000001b3L;
            value = (value ^ track.fingerprint.hashCode()) * 0x100000001b3L;
        }
        return value;
    }

    private void notifyUi() {
        synchronized (this) {
            if (updatePosted || closed) {
                return;
            }
            updatePosted = true;
        }
        host.uiHandler.postDelayed(() -> {
            synchronized (SoundAnalysisController.this) {
                updatePosted = false;
            }
            if (!host.isFinishing()
                    && host.navigationState.tabIndex == LibraryTabs.SOUND) {
                host.render();
            }
            host.refreshSettingsLabels();
        }, 200L);
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/SoundAnalysisController.kt
package com.dumuzeyn.mp3player

import android.content.Context
import android.os.Process
import java.io.Closeable
import java.util.LinkedHashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicInteger

/** Runs one resumable, local-only analysis queue for the current process. */
internal class SoundAnalysisController(private val host: MainActivityCore) : Closeable {
    private var store: SoundProfileStore? = null
    private var extractor: AudioFeatureExtractor? = null
    private val clusterEngine = SoundClusterEngine()
    private val executor: ExecutorService = Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "voltune-sound-analysis").apply { priority = Thread.MIN_PRIORITY }
    }
    private val generation = AtomicInteger()

    @Volatile private var groups = ArrayList<SoundGroup>()
    @Volatile private var total = 0
    @Volatile private var analyzed = 0
    @Volatile private var failed = 0
    @Volatile private var queued = 0
    @Volatile private var activeTitle = ""
    @Volatile private var rebuildingGroups = false
    @Volatile private var fullReanalysis = false
    @Volatile private var blockReason = SoundAnalysisConstraints.BlockReason.NONE
    @Volatile private var closed = false
    @Volatile private var lastLibrarySignature = Long.MIN_VALUE
    private var updatePosted = false

    fun enabled(): Boolean = host.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        .getBoolean(ENABLED, true)

    fun toggle() {
        val value = !enabled()
        host.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putBoolean(ENABLED, value)
            .apply()
        if (value) {
            lastLibrarySignature = Long.MIN_VALUE
            onLibraryReady(host.libraryState.tracks)
        } else {
            generation.incrementAndGet()
            activeTitle = ""
            rebuildingGroups = false
            fullReanalysis = false
            blockReason = SoundAnalysisConstraints.BlockReason.NONE
            notifyUi()
        }
    }

    fun settingLabel(): String {
        val isEnabled = enabled()
        return host.tr("Analyze similar tracks: ", "Анализ похожих треков: ") +
            host.tr(if (isEnabled) "on" else "off", if (isEnabled) "вкл" else "выкл")
    }

    fun onLibraryReady(tracks: List<Track>) {
        if (closed) return
        val signature = librarySignature(tracks)
        if (signature == lastLibrarySignature) return
        lastLibrarySignature = signature
        val requestedGeneration = generation.incrementAndGet()
        val snapshot = ArrayList(tracks)
        executor.execute { runQueueSafely(requestedGeneration, snapshot) }
    }

    fun groups(): ArrayList<SoundGroup> = ArrayList(groups)

    fun total(): Int = total

    fun analyzed(): Int = analyzed

    fun failed(): Int = failed

    fun queued(): Int = queued

    fun activeTitle(): String = activeTitle

    fun rebuildingGroups(): Boolean = rebuildingGroups

    fun fullReanalysis(): Boolean = fullReanalysis

    @Synchronized
    fun rebuildGroupsFromSavedProfiles(): Boolean {
        if (closed || rebuildingGroups || fullReanalysis) return false
        val requestedGeneration = generation.incrementAndGet()
        rebuildingGroups = true
        activeTitle = ""
        notifyUi()
        executor.execute {
            try {
                ensureWorkerResources()
                val profiles = requireNotNull(store).loadProfiles()
                if (requestedGeneration == generation.get()) rebuildGroups(profiles)
            } catch (error: RuntimeException) {
                VoltuneLog.failure("sound_group_rebuild_failed", error)
            } finally {
                rebuildingGroups = false
                notifyUi()
            }
        }
        return true
    }

    @Synchronized
    fun reanalyzeLibrary(): Boolean {
        if (closed || rebuildingGroups || fullReanalysis) return false
        host.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putBoolean(ENABLED, true)
            .apply()
        val requestedGeneration = generation.incrementAndGet()
        val snapshot = ArrayList(host.libraryState.tracks)
        fullReanalysis = true
        total = snapshot.size
        analyzed = 0
        failed = 0
        queued = snapshot.size
        activeTitle = ""
        notifyUi()
        executor.execute {
            try {
                ensureWorkerResources()
                requireNotNull(store).clearAnalysis()
                groups = ArrayList()
                lastLibrarySignature = librarySignature(snapshot)
                runQueueSafely(requestedGeneration, snapshot)
            } catch (error: RuntimeException) {
                VoltuneLog.failure("sound_full_reanalysis_failed", error)
                fullReanalysis = false
                notifyUi()
            } finally {
                if (queued == 0 || requestedGeneration != generation.get()) {
                    fullReanalysis = false
                    notifyUi()
                }
            }
        }
        return true
    }

    fun blockReason(): SoundAnalysisConstraints.BlockReason = blockReason

    override fun close() {
        if (closed) return
        closed = true
        rebuildingGroups = false
        fullReanalysis = false
        generation.incrementAndGet()
        executor.execute(::closeWorkerResources)
        executor.shutdown()
    }

    private fun runQueueSafely(requestedGeneration: Int, tracks: ArrayList<Track>) {
        try {
            runQueue(requestedGeneration, tracks)
        } catch (error: RuntimeException) {
            VoltuneLog.failure("sound_analysis_queue_failed", error)
            closeWorkerResources()
            activeTitle = ""
            rebuildingGroups = false
            fullReanalysis = false
            blockReason = SoundAnalysisConstraints.BlockReason.NONE
            if (closed || requestedGeneration != generation.get()) return
            lastLibrarySignature = Long.MIN_VALUE
            notifyUi()
            host.uiHandler.postDelayed(
                {
                    if (!closed && requestedGeneration == generation.get()) {
                        onLibraryReady(host.libraryState.tracks)
                    }
                },
                CONSTRAINT_RECHECK_MS,
            )
        }
    }

    private fun closeWorkerResources() {
        store?.let { profileStore ->
            try {
                profileStore.close()
            } catch (error: RuntimeException) {
                VoltuneLog.failure("sound_profile_store_close_failed", error)
            }
            store = null
        }
        extractor = null
    }

    private fun runQueue(requestedGeneration: Int, tracks: ArrayList<Track>) {
        Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND)
        if (!isCurrent(requestedGeneration)) return
        ensureWorkerResources()
        val profileStore = requireNotNull(store)
        val featureExtractor = requireNotNull(extractor)
        val profiles = profileStore.loadProfiles()
        groups = profileStore.loadGroups()
        val pending = prepareQueue(tracks, profiles)
        val pendingCount = pending.size
        var clusteringChanged = host.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getInt(CLUSTERING_VERSION, 0) != SoundClusterEngine.CLUSTERING_VERSION
        profileStore.pruneEmptyGroups()
        groups = profileStore.loadGroups()
        publishCounts(tracks.size, profiles, pending.size)
        notifyUi()
        if (!enabled()) return
        if (clusteringChanged && pending.isEmpty() && usableCount(profiles) >= 4) {
            rebuildGroups(profiles)
            clusteringChanged = false
        }
        for (track in pending) {
            if (!runAllowedOrRescheduled(requestedGeneration, tracks)) return
            activeTitle = track.title
            profileStore.mark(track, SoundAnalysisState.ANALYZING, "")
            notifyUi()
            try {
                val features = featureExtractor.analyze(track) {
                    shouldYield(requestedGeneration)
                }
                if (!isCurrent(requestedGeneration)) return
                val profile = TrackAudioProfile.analyzed(track, features)
                profileStore.saveProfile(profile)
                profiles[track.trackId] = profile
                analyzed++
                queued = maxOf(0, queued - 1)
                updateGroupsAfterProfile(profiles, profile)
            } catch (_: AudioFeatureExtractor.AnalysisInterruptedException) {
                if (!isCurrent(requestedGeneration)) return
                profileStore.mark(track, SoundAnalysisState.QUEUED, "")
                lastLibrarySignature = Long.MIN_VALUE
                onLibraryReady(tracks)
                return
            } catch (error: Exception) {
                if (!isCurrent(requestedGeneration)) return
                val errorName = error.javaClass.simpleName
                profileStore.mark(track, SoundAnalysisState.FAILED, errorName)
                profiles[track.trackId] = TrackAudioProfile(
                    track.trackId,
                    TrackAudioProfile.ANALYSIS_VERSION,
                    track.fileSize,
                    track.lastModified,
                    track.fingerprint,
                    SoundAnalysisState.FAILED,
                    DoubleArray(0),
                    "",
                    errorName,
                    System.currentTimeMillis(),
                )
                failed++
                queued = maxOf(0, queued - 1)
                VoltuneLog.failure("sound_analysis_failed", error)
            }
            activeTitle = ""
            notifyUi()
        }
        if (pendingCount > 0 || clusteringChanged || groups.isEmpty()) rebuildGroups(profiles)
        activeTitle = ""
        blockReason = SoundAnalysisConstraints.BlockReason.NONE
        fullReanalysis = false
        notifyUi()
    }

    private fun ensureWorkerResources() {
        if (store == null) {
            store = SoundProfileStore(host)
            extractor = AudioFeatureExtractor(host)
        }
    }

    private fun prepareQueue(
        tracks: ArrayList<Track>,
        profiles: LinkedHashMap<String, TrackAudioProfile>,
    ): ArrayList<Track> {
        val profileStore = requireNotNull(store)
        val pending = ArrayList<Track>()
        for (track in tracks) {
            val profile = profiles[track.trackId]
            if (profile != null && profile.matches(track) &&
                (profile.usable() || profile.state == SoundAnalysisState.FAILED)
            ) {
                continue
            }
            profileStore.mark(track, SoundAnalysisState.QUEUED, "")
            profiles[track.trackId] = TrackAudioProfile.pending(track, SoundAnalysisState.QUEUED)
            pending.add(track)
        }
        return pending
    }

    private fun updateGroupsAfterProfile(
        profiles: Map<String, TrackAudioProfile>,
        profile: TrackAudioProfile,
    ) {
        if (groups.isEmpty() && usableCount(profiles) >= 4) {
            rebuildGroups(profiles)
            return
        }
        if (groups.isNotEmpty()) {
            val groupId = clusterEngine.nearestGroup(
                profile.features,
                ArrayList(profiles.values),
                groups,
            )
            requireNotNull(store).assign(profile.trackId, groupId)
            groups = requireNotNull(store).loadGroups()
        }
        if (usableCount(profiles) <= 20) rebuildGroups(profiles)
    }

    private fun rebuildGroups(profiles: Map<String, TrackAudioProfile>) {
        val profileStore = requireNotNull(store)
        val rebuilt = clusterEngine.cluster(ArrayList(profiles.values))
        profileStore.replaceGroups(rebuilt)
        groups = profileStore.loadGroups()
        host.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putInt(CLUSTERING_VERSION, SoundClusterEngine.CLUSTERING_VERSION)
            .apply()
    }

    private fun runAllowedOrRescheduled(
        requestedGeneration: Int,
        tracks: ArrayList<Track>,
    ): Boolean {
        if (!isCurrent(requestedGeneration)) return false
        blockReason = SoundAnalysisConstraints.reason(host, host.isPlaybackPlaying())
        if (blockReason == SoundAnalysisConstraints.BlockReason.NONE) return true
        activeTitle = ""
        notifyUi()
        host.uiHandler.postDelayed(
            {
                if (!closed && enabled() && requestedGeneration == generation.get()) {
                    lastLibrarySignature = Long.MIN_VALUE
                    onLibraryReady(tracks)
                }
            },
            CONSTRAINT_RECHECK_MS,
        )
        return false
    }

    private fun shouldYield(requestedGeneration: Int): Boolean =
        !isCurrent(requestedGeneration) ||
            SoundAnalysisConstraints.reason(host, host.isPlaybackPlaying()) !=
            SoundAnalysisConstraints.BlockReason.NONE

    private fun isCurrent(requestedGeneration: Int): Boolean =
        !closed && enabled() && requestedGeneration == generation.get() &&
            !Thread.currentThread().isInterrupted

    private fun publishCounts(
        trackCount: Int,
        profiles: Map<String, TrackAudioProfile>,
        pendingCount: Int,
    ) {
        total = trackCount
        analyzed = 0
        failed = 0
        for (profile in profiles.values) {
            if (profile.usable()) {
                analyzed++
            } else if (profile.state == SoundAnalysisState.FAILED) {
                failed++
            }
        }
        queued = pendingCount
    }

    private fun notifyUi() {
        synchronized(this) {
            if (updatePosted || closed) return
            updatePosted = true
        }
        host.uiHandler.postDelayed(
            {
                synchronized(this) { updatePosted = false }
                if (!host.isFinishing && host.navigationState.tabIndex == LibraryTabs.SOUND) {
                    host.render()
                }
                host.refreshSettingsLabels()
            },
            200L,
        )
    }

    companion object {
        private const val PREFS = "mp3_player_ui"
        private const val ENABLED = "soundAnalysisEnabled"
        private const val CLUSTERING_VERSION = "similarClusteringVersion"
        private const val CONSTRAINT_RECHECK_MS = 5_000L

        private fun usableCount(profiles: Map<String, TrackAudioProfile>): Int =
            profiles.values.count(TrackAudioProfile::usable)

        private fun librarySignature(tracks: List<Track>): Long {
            var value = 0xcbf29ce484222325UL.toLong()
            for (track in tracks) {
                value = (value xor track.trackId.hashCode().toLong()) * 0x100000001b3L
                value = (value xor track.fileSize) * 0x100000001b3L
                value = (value xor track.lastModified) * 0x100000001b3L
                value = (value xor track.fingerprint.hashCode().toLong()) * 0x100000001b3L
            }
            return value
        }
    }
}
